File: java/src/main/java/problems101_200/Triangle.java
package problems101_200;

import java.util.ArrayList;
import java.util.List;

enum Triangle {
    DEFAULT {
        @Override
        int minimumTotal(List<List<Integer>> triangle) {
            int maxRow = triangle.size();
            int maxCol = 2;

            List<List<Integer>> dp = new ArrayList<>(maxRow);
            for (int i = 0; i < maxRow; i++) {
                dp.add(new ArrayList<>());
            }
            dp.get(0).add(triangle.get(0).get(0));
            int row = 1;
            int col = 0;
            while (row < maxRow) {
                List<Integer> columns = new ArrayList<>();
                if (col > 0) {
                    columns.add(triangle.get(row).get(col) + dp.get(row - 1).get(col - 1));
                }
                if (col < maxCol - 1) {
                    columns.add(triangle.get(row).get(col) + dp.get(row - 1).get(col));
                }
                dp.get(row).add(getMinimum(columns));
                col++;
                if (col != maxCol) {
                    continue;
                }
                col = 0;
                row++;
                maxCol++;
            }
            return getMinimum(dp.get(dp.size() - 1));
        }

        int getMinimum(List<Integer> columns) {
            int minimum = columns.get(0);
            for (int i = 1; i < columns.size(); i++) {
                if (columns.get(i) < minimum) {
                    minimum = columns.get(i);
                }
            }
            return minimum;
        }
    };

    abstract int minimumTotal(List<List<Integer>> triangle);
}

File: kotlin/src/main/kotlin/problems101_200/Triangle.kt
package problems101_200

enum class Triangle {
    DEFAULT {
        override fun minimumTotal(triangle: List<List<Int>>): Int {
            val maxRow = triangle.size
            var maxCol = 2

            val dp = ArrayList<MutableList<Int>>(maxRow)
            for (i in 0 until maxRow) {
                dp.add(ArrayList())
            }
            dp[0].add(triangle[0][0])
            var row = 1
            var col = 0

            while (row < maxRow) {
                val columns = arrayListOf<Int>()
                if (col > 0) {
                    columns += triangle[row][col] + dp[row - 1][col - 1]
                }
                if (col < maxCol - 1) {
                    columns += triangle[row][col] + dp[row - 1][col]
                }
                dp[row] += getMinimum(columns)
                col++
                if (col != maxCol) {
                    continue
                }
                col = 0
                row++
                maxCol++
            }
            return getMinimum(dp[dp.size - 1])
        }

        fun getMinimum(columns: List<Int>): Int {
            var minimum = columns[0]
            for (i in 1 until columns.size) {
                if (columns[i] < minimum) {
                    minimum = columns[i]
                }
            }
            return minimum
        }
    },
    ;

    abstract fun minimumTotal(triangle: List<List<Int>>): Int
}
